### DISCUSSION TITLE
How to filter by content of files within selected files or list of files showing

### DISCUSSION THREAD

**Discussion by wjdwndud0114** [ibhagwan/fzf-lua#1188] [Q&A] [answered]
Hi, I have custom rg/fd functions like this (one example):
```
  vim.keymap.set('n', '<leader><C-p>', function()
    fzf.fzf_exec('fd -H --type f --strip-cwd-prefix | ~/.dotfiles/file-web-devicon', {
      actions = fzf.defaults.actions.files,
      cwd = vim.api.nvim_eval("expand('%:p:~:.:h')"),
      prompt = vim.api.nvim_eval("expand('%:p:~:.:h')") .. '> ',
      fzf_opts = {
        ['--multi'] = 999,
        ['--nth'] = 2,
        ['--delimiter'] = fzf.utils.nbsp
      },
      previewer = 'builtin',
    })
  end, { noremap = true })
```
The purpose being to replace the icon generating logic with something faster which is essential with enormous monorepos.

This lets me find files within the current parent directory.

<img width="1439" alt="image" src="https://github.com/ibhagwan/fzf-lua/assets/30526747/a09129dc-5d37-4ccb-9d5e-1333c3eec418">

However, I often need to do an additional search within the files that were searched. Is there a way to "rg" for content insides these files? Looking for suggestions for a solution.

I'm thinking I would have to implement a kepmap that would get the list of current files or selections from fzf-lua and then open a live grep window filtered with the given files?

I know there is a `<C-g>` with the builtin `live_grep` but I can't use that due to it being too slow (icons slow without the custom icon calculation). If it's possible to change my setup to use this as well as custom icons, please let me know!

Thank you!

**Comment by ibhagwan** [comment]
> I'm thinking I would have to implement a kepmap that would get the list of current files or selections from fzf-lua and then open a live grep window filtered with the given files?

Correct, you’d need to construct a special rg command from the list of files and feed that into `live_grep({ cmd = … })`.

> I know there is a with the builtin live_grep but I can't use that due to it being too slow (icons slow without the custom icon calculation). If it's possible to change my setup to use this as well as custom icons, please let me know!

Not sure if I understand this question, you can always use `file_icons=false,git_icons=false` if you feel the overhead is too big.

**Comment by ibhagwan** [comment, marked as answer]
I missed the part where your icon generating logic is actually in the shell:
```sh
fd -H --type f --strip-cwd-prefix | ~/.dotfiles/file-web-devicon
```

> I know there is a with the builtin live_grep but I can't use that due to it being too slow (icons slow without the custom icon calculation). If it's possible to change my setup to use this as well as custom icons, please let me know!

Why don’t you call `fzf_live` with a custom `rg` command piped into your iconifier in the same way you called `fzf_exec`?
> Assuming your iconifier can handle `file:line:col:text` output from `rg`

```lua
fzf.fzf_live(
    "rg --line-number --column --color=always -- <query> | ~/.dotfiles/file-web-devicon", {
      actions = fzf.defaults.actions.files,
      cwd = vim.api.nvim_eval("expand('%:p:~:.:h')"),
      prompt = vim.api.nvim_eval("expand('%:p:~:.:h')") .. '> ',
      fzf_opts = {
        ['--multi'] = 999,
        ['--nth'] = 2,
        ['--delimiter'] = fzf.utils.nbsp
      },
      previewer = 'builtin',
  })
```

**Reply by wjdwndud0114** [reply]
Thanks for the response!

I do have that already for `rg` but that doesn't come with `<C-g>` functionality
```
  vim.keymap.set('n', '<leader><leader>s', function()
    fzf.fzf_live(
      'rg --column --line-number --no-heading --color=always --smart-case -- <query> | ~/.dotfiles/file-web-devicon', {
        actions = fzf.defaults.actions.files,
        cwd = vim.api.nvim_eval("expand('%:p:~:.:h')"),
        prompt = vim.api.nvim_eval("expand('%:p:~:.:h')") .. ' Rg> ',
        fzf_opts = {
          ['--multi'] = 999,
          ['--nth'] = 2,
          ['--delimiter'] = fzf.utils.nbsp
        },
        previewer = 'builtin',
      })
  end, { noremap = true })
```

<img width="1075" alt="image" src="https://github.com/ibhagwan/fzf-lua/assets/30526747/20f9366d-b6de-46ea-939a-18be4f1d8352">

**Reply by wjdwndud0114** [reply]
Or maybe I should just disable icons and use live_grep, git_files 🤔

**Reply by wjdwndud0114** [reply]
Although, it would be very useful to be able to `rg` within files showing such as in `oldfiles`, `gitfiles`, etc...

**Reply by ibhagwan** [reply]
> I do have that already for rg but that doesn't come with `<C-g>` functionality

It is possible but that would require creating your own version of the function switching between your custom fzf_exec / fzf_live, you can take a look at the action to see how it’s done.

SAlternatively it’s possible to reuse `grep|live_grep`, use the `opts.filter` field as your icon generator and nullify the make entry processor to avoid adding the icons twice with  `__mt_transform = 'return function(x) return x end'`, you can explore the code for `tags` to see how grep is reused:
https://github.com/ibhagwan/fzf-lua/blob/9c953dfa7650191d892800333d187f3439e01afd/lua/fzf-lua/providers/tags.lua#L152-L175

> Or maybe I should just disable icons and use live_grep, git_files

Certainly an option :)

How much faster is your script anyways (vs the lua code)?

> Although, it would be very useful to be able to rg within files showing such as in oldfiles, gitfiles, etc...

These seems like a different issue/feature unrelated to the “faster icons”, as I mentioned, this would require crafting your own rg command from the selected files (searching only in these) and feeding that command as `cmd`.

**Reply by wjdwndud0114** [reply]
Thank you, will explore and see what I can do when I have time.

> How much faster is your script anyways (vs the lua code)?

Hmm just tested it and around ~2seconds faster to load although I remember the difference being much greater. Maybe something changed since then :) This is for 500k+ files. ([this is the binary](https://github.com/MaartenStaa/file-web-devicons))

**Reply by ibhagwan** [reply]
2s shouldn't make such a difference for such a large project, is it possible the difference was due to `git_icons=true` (which requires running `git status` for the git indicators)?

I've since added a warning to alert users if there's a `git status` which is taking more than 3s:
https://github.com/ibhagwan/fzf-lua/blob/7109a18dd1832703209fc8f46e85df3c6ed6d059/lua/fzf-lua/make_entry.lua#L79-L80

> https://github.com/MaartenStaa/file-web-devicons

Nice project! in the past I looked at https://github.com/coreyja/devicon-lookup but the benefit wasn't worth the hassle (having to supply a precompiled binary with fzf-lua) but you might find it interesting as well as [the writeup](https://coreyja.com/posts/vim-fzf-with-devicons).

**Comment by ibhagwan** [comment]
@wjdwndud0114, FYI, ty for letting me know this existed, might be useful one day :-)

https://github.com/MaartenStaa/file-web-devicons/issues/4
